embassy-stm32/src/dma/dma.rs

```diff
@@ -56,6 +57,93 @@ impl State {
 
 static STATE: State = State::new();
 
+struct M0AR(u32);
+struct M1AR(u32);
+struct ChunkSize(u16);
+
+struct GiantTransferState {
+    enabled: [AtomicBool; DMA_CHANNEL_COUNT],
+    remaining_chunks: [AtomicU32; DMA_CHANNEL_COUNT],
+    chunk_size: [AtomicU16; DMA_CHANNEL_COUNT],
+    transfer_len_bytes: [AtomicU8; DMA_CHANNEL_COUNT],
+}
+
+impl GiantTransferState {
+    const fn new() -> Self {
+        const EN: AtomicBool = AtomicBool::new(false);
+        const RC: AtomicU32 = AtomicU32::new(0);
+        const CS: AtomicU16 = AtomicU16::new(0);
+        const DL: AtomicU8 = AtomicU8::new(0);
+        Self {
+            enabled: [EN; DMA_CHANNEL_COUNT],
+            remaining_chunks: [RC; DMA_CHANNEL_COUNT],
+            chunk_size: [CS; DMA_CHANNEL_COUNT],
+            transfer_len_bytes: [DL; DMA_CHANNEL_COUNT],
+        }
+    }
+
+    fn enable_for_channel_index(
+        &self,
+        channel_index: usize,
+        data_addr: u32,
+        data_len: usize,
+        transfer_len_bytes: u8,
+    ) -> (M0AR, M1AR, ChunkSize) {
+        let mut chunk_size = data_len;
+        let mut total_chunk_count = 1;
+        while chunk_size > 0xffff {
+            chunk_size /= 2;
```

Comment:
this is wrong if `data_len` doesn't divide evenly (for example if it's `2^17 + 1`), you're rounding down so it'll read less bytes than wanted.

Reply:
You are right, the giant transfer cannot be used wit odd data length, as there is no way of changing the data length without any gap between the transfers.

The Channel trait impl checks for this case and panics, I also added some asserts that checks that the data len is even.

Reply:
It's not enough to check it's even. If you do 4 chunks it has to be multiple of 4, if 8 chunks it has to be multiple of 8, etc...

Reply:
You are right, maybe dcmi buffers lengths uphold some properties that makes it work for them. https://github.com/STMicroelectronics/STM32CubeH7/blob/b340b13929e36a3427b8d94e8b1006022f82273f/Drivers/STM32H7xx_HAL_Driver/Src/stm32h7xx_hal_dcmi.c#L419 I’ll do some calculations soon to see if it isn’t a bug in cube as well.